avahi-daemon/static-services.c

```diff
@@ -420,6 +449,121 @@ static void XMLCALL xml_start(void *data, const char *el, const char *attr[]) {
     return;
 }
 
+static uint8_t hex (char c)
+{
+  if ((c >= '0') && (c <= '9'))
+    return c - '0';
+  if ((c >= 'A') && (c <= 'F'))
+    return c - 'A' + 10;
+  if ((c >= 'a') && (c <= 'f'))
+    return c - 'a' + 10;
+  return 0xFF;
+}
+
+static int decode_hex_buf (struct xml_userdata *u, uint8_t **out_buf, size_t *out_buf_len) {
+    size_t buf_len = strlen(u->buf);
+    uint8_t *raw_buf;
+    size_t iter;
+    size_t raw_buf_len;
+
+    if (buf_len % 2) {
+        avahi_log_error("%s: parse failure: hex value of the txt record should have an even length", u->group->filename);
+        u->failed = 1;
+        return -1;
+    }
+    raw_buf_len = buf_len / 2;
+    raw_buf = avahi_malloc(raw_buf_len);
+    for (iter = 0; iter < raw_buf_len; iter++) {
+        uint8_t high_nibble = hex (u->buf[iter * 2]);
```

Comment:
While replacing that I also spotted one bug, where a free a buffer with `free` instead of `avahi_free`. Fixed that one too.